### PR TITLE
Info add swarm property.

### PR DESCRIPTION
**Pull request by hatunet**
#733 
now, info interface added a new property swarm.

### CODE REVIEW

File: src/main/java/com/spotify/docker/client/messages/Info.java [hunk 1]
```diff
@@ -259,7 +267,8 @@ static Info create(
         httpProxy, httpsProxy, id, ipv4Forwarding, images, indexServerAddress, initPath, initSha1,
         kernelMemory, kernelVersion, labelsT, memTotal, memoryLimit, cpus, eventsListener,
         fileDescriptors, goroutines, name, noProxy, oomKillDisable, operatingSystem, osType,
-        plugins, registryConfig, serverVersion, swapLimit, systemStatusB.build(), systemTime);
+        plugins, registryConfig, serverVersion, swapLimit, systemStatusB.build(), systemTime,
+            swarm);
```
Comment (johnflavin) [inline src/main/java/com/spotify/docker/client/messages/Info.java]: This line should be indented with the line above.

File: src/main/java/com/spotify/docker/client/messages/Info.java [hunk 2]
```diff
@@ -332,4 +341,251 @@ static IndexConfig create(
...
+            @JsonProperty("NodeID") String nodeId,
+            @JsonProperty("NodeAddr") String nodeAddr,
+            @JsonProperty("LocalNodeState") String localNodeState,
+            @JsonProperty("ControlAvailable") boolean controlAvailable,
+            @JsonProperty("Error") boolean error,
+            @JsonProperty("RemoteManagers") ImmutableList<RemoteManager> remoteManagers,
+            @JsonProperty("Nodes") int nodes,
+            @JsonProperty("Managers") int managers,
+            @JsonProperty("Cluster") Cluster cluster
+    ) {
```
Comment (johnflavin) [inline src/main/java/com/spotify/docker/client/messages/Info.java]: Close the parentheses and open the curly braces on the previous line.

File: src/main/java/com/spotify/docker/client/messages/Info.java [hunk 3]
```diff
@@ -332,4 +341,251 @@ static IndexConfig create(
...
+    public abstract SwarmSpec spec();
+
+    @JsonCreator
+    static Cluster create(
+            @JsonProperty("ID") String id,
+            @JsonProperty("Version") Version version,
+            @JsonProperty("CreatedAt") Date createdAt,
+            @JsonProperty("UpdatedAt") Date updatedAt,
+            @JsonProperty("Spec") SwarmSpec spec
+    ) {
```
Comment (johnflavin) [inline src/main/java/com/spotify/docker/client/messages/Info.java]: Close the parentheses and open the curly braces on the previous line.

**Comment by hatunet** [comment]
@johnflavin 
https://travis-ci.org/spotify/docker-client/jobs/242072661 
always wrong

**Comment by johnflavin** [comment]
Yeah, we see that a lot. I have no idea why it happens,  but sometimes the docker daemon on the version 1.10 tests just fails to connect, and a bunch of the tests fail for no good reason. It's one of the reasons we want to move more to unit testing.

Don't worry, though. We understand that it isn't the fault of your code.

File: src/test/java/com/spotify/docker/client/DefaultDockerClientTest.java [hunk 4]
```diff
@@ -777,6 +777,7 @@ public void testInfo() throws Exception {
     if (dockerApiVersionAtLeast("1.23")) {
       assertThat(info.cgroupDriver(), not(isEmptyOrNullString()));
       assertThat(info.kernelMemory(), is(anything()));
+      assertThat(info.swarm(), is(anything()));
```
Comment (davidxia) [inline src/test/java/com/spotify/docker/client/DefaultDockerClientTest.java]: Is it easy to write an integration test that actually checks the values here?

File: src/main/java/com/spotify/docker/client/messages/Info.java [hunk 5]
```diff
@@ -332,4 +341,249 @@ static IndexConfig create(
       return new AutoValue_Info_IndexConfig(name, mirrorsT, secure, official);
     }
   }
+
+  @AutoValue
```
Comment (davidxia) [inline src/main/java/com/spotify/docker/client/messages/Info.java]: Why do you add `@JsonAutoDetect` for some inner classes but not others?

File: src/main/java/com/spotify/docker/client/messages/Info.java [hunk 6]
```diff
@@ -332,4 +341,249 @@ static IndexConfig create(
...
+            @JsonProperty("Version") Version version,
+            @JsonProperty("CreatedAt") Date createdAt,
+            @JsonProperty("UpdatedAt") Date updatedAt,
+            @JsonProperty("Spec") SwarmSpec spec) {
+      return new AutoValue_Info_Cluster(id, version, createdAt, updatedAt, spec);
+    }
+  }
+
+  @AutoValue
+  public abstract static class Version {
```
Comment (davidxia) [inline src/main/java/com/spotify/docker/client/messages/Info.java]: Any reason not to use `com.spotify.docker.client.messages.swarm.Version`?

File: src/main/java/com/spotify/docker/client/messages/Info.java [hunk 7]
```diff
@@ -332,4 +341,249 @@ static IndexConfig create(
...
+    public abstract Integer index();
+
+    @JsonCreator
+    static Version create(@JsonProperty("ID") Integer index) {
+      return new AutoValue_Info_Version(index);
+    }
+  }
+
+  @AutoValue
+  public abstract static class SwarmSpec {
```
Comment (davidxia) [inline src/main/java/com/spotify/docker/client/messages/Info.java]: Any reason to not use `com.spotify.docker.client.messages.swarm.SwarmSpec`?

File: src/main/java/com/spotify/docker/client/messages/Info.java [hunk 8]
```diff
@@ -332,4 +341,249 @@ static IndexConfig create(
...
+      final ImmutableMap<String, String> labelsT = labels == null
+              ? null : ImmutableMap.copyOf(labels);
+      return new AutoValue_Info_SwarmSpec(name, labelsT, orchestration, raft, dispatcher,
+              caConfig, taskDefaults);
+    }
+  }
+
+  @AutoValue
+  @JsonAutoDetect(fieldVisibility = ANY, getterVisibility = NONE, setterVisibility = NONE)
+  public abstract static class OrchestrationConfig {
```
Comment (davidxia) [inline src/main/java/com/spotify/docker/client/messages/Info.java]: There's already a `com.spotify.docker.client.messages.swarm.OrchestrationConfig`.

File: src/main/java/com/spotify/docker/client/messages/Info.java [hunk 9]
```diff
@@ -332,4 +341,249 @@ static IndexConfig create(
...
+    @JsonCreator
+    static OrchestrationConfig create(
+            @JsonProperty("TaskHistoryRetentionLimit") final Integer taskHistoryRetentionLimit) {
+      return new AutoValue_Info_OrchestrationConfig(taskHistoryRetentionLimit);
+    }
+  }
+
+  @AutoValue
+  @JsonAutoDetect(fieldVisibility = ANY, getterVisibility = NONE, setterVisibility = NONE)
+  public abstract static class RaftConfig {
```
Comment (davidxia) [inline src/main/java/com/spotify/docker/client/messages/Info.java]: There's already `com.spotify.docker.client.messages.swarm.RaftConfig`.

File: src/main/java/com/spotify/docker/client/messages/Info.java [hunk 10]
```diff
@@ -332,4 +341,249 @@ static IndexConfig create(
...
+            @JsonProperty("ElectionTick") final Integer electionTick,
+            @JsonProperty("HeartbeatTick") final Integer heartbeatTick) {
+      return new AutoValue_Info_RaftConfig(snapshotInterval, keepOldSnapshots,
+              logEntriesForSlowFollowers, electionTick, heartbeatTick);
+    }
+  }
+
+  @AutoValue
+  @JsonAutoDetect(fieldVisibility = ANY, getterVisibility = NONE, setterVisibility = NONE)
+  public abstract static class DispatcherConfig {
```
Comment (davidxia) [inline src/main/java/com/spotify/docker/client/messages/Info.java]: See `com.spotify.docker.client.messages.swarm.DispatcherConfig`

File: src/main/java/com/spotify/docker/client/messages/Info.java [hunk 11]
```diff
@@ -332,4 +341,249 @@ static IndexConfig create(
...
+
+    @JsonCreator
+    static DispatcherConfig create(@JsonProperty("HeartbeatPeriod") final Long heartbeatPeriod) {
+      return new AutoValue_Info_DispatcherConfig(heartbeatPeriod);
+    }
+  }
+
+  @AutoValue
+  @JsonAutoDetect(fieldVisibility = ANY, getterVisibility = NONE, setterVisibility = NONE)
+  public abstract static class CaConfig {
```
Comment (davidxia) [inline src/main/java/com/spotify/docker/client/messages/Info.java]: already a class for this

**Comment by davidxia** [comment]
Right now, I prefer the approach of https://github.com/spotify/docker-client/pull/754, but the author seems to not have followed up. I'd be alright with something more like https://github.com/spotify/docker-client/pull/754 given that the things in that PR are addressed.

**Comment by hatunet** [comment]
hi @davidxia ,
I will fix this PR's code  follow your review suggestion.

**Comment by davidxia** [comment]
Closed in favor of https://github.com/spotify/docker-client/pull/814

**Pull request closed without merging**